contracts/governance/ApprovalReceiver.sol

```diff
@@ -18,12 +18,39 @@ contract ApprovalReceiver is ErrorDecoder, IApproveAndCall {
 	 * @param _data the data will be used for low level call
 	 */
 	function receiveApproval(
-		address, /*_sender*/
-		uint256, /*_amount*/
-		address, /*_token*/
-		bytes memory _data
-	) public {
-		_receiveApproval(_getToken(), _data, _getSelectors());
+		address _sender,
+		uint256 _amount,
+		address _token,
+		bytes calldata _data
+	) external {
+		//accepts calls only from SOV token
+		require(msg.sender == _getToken(), "unauthorized");
+		require(msg.sender == _token, "unauthorized");
+
+		//only allowed methods
+		bool isAllowed = false;
+		bytes4[] memory selectors = _getSelectors();
+		bytes4 sig = _getSig(_data);
+		for (uint256 i = 0; i < selectors.length; i++) {
+			if (sig == selectors[i]) {
+				isAllowed = true;
+				break;
+			}
+		}
+		require(isAllowed, "method is not allowed");
+
+		//check sender and amount
+		bytes32 dummy;
+		address sender;
+		uint256 amount;
+		(dummy, sender, amount) = abi.decode(
```

Comment:
jfyi: `(, address sender, uint256 amount) = abi.decode(...` - a more compact notation - I've skipped `dummy` and linlined variable declaration

Reply:
agree :-)